aten/src/ATen/core/Tensor.h

```diff
@@ -547,6 +547,9 @@ class CAFFE2_API Tensor {
   Tensor std(IntArrayRef dim, bool unbiased=true, bool keepdim=false) const;
   Tensor prod(c10::optional<ScalarType> dtype=c10::nullopt) const;
   Tensor prod(int64_t dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
+  #ifdef BUILD_NAMEDTENSOR
+  Tensor prod(Dimname dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
```

Comment:
hmm... any thoughts on making this kwonly instead? we could use a `#else` to define the alternate interface. That way we'd have fewer interfaces to support long-term, and less risk of eventually changing something in a way that creates the potential for ambiguous calls.

Reply:
or maybe we could just add kw-only dimname args to the existing calls and have the implementation throw `#if ! defined` ... though I guess that means the `Dimname` class has to be present.

Reply:
I'm not sure what you mean by "making this kwonly" -- do you mean moving `keepdim, dtype` to be kwonly args? This doesn't fix the problem that we'd have two schema strings. 

One potential solution long-term that I've been looking at is creating a `Dim` union type: `Dim = union[Dimname, int64_t]` and merging the two schemas. This isn't in my plans for the first iteration of named tensor but having something like it would definitely fix the problem of supporting too many interfaces.

Reply:
oh nevermind, I wasn't thinking clearly about the behavior of this function.